spec/datadog/tracing/contrib/pg/patcher_spec.rb

```diff
@@ -0,0 +1,788 @@
+
+        it 'produces a trace' do
+          expect(spans.count).to eq(1)
+          expect(span.name).to eq(Datadog::Tracing::Contrib::Pg::Ext::SPAN_SYNC_EXEC_PREPARED)
+          expect(span.resource).to eq(statement_name)
+          expect(span.service).to eq('pg')
+          expect(span.type).to eq(Datadog::Tracing::Metadata::Ext::SQL::TYPE)
+          expect(span.get_tag(Datadog::Tracing::Metadata::Ext::TAG_KIND))
+            .to eq(Datadog::Tracing::Metadata::Ext::SpanKind::TAG_CLIENT)
+          expect(span.get_tag(Datadog::Tracing::Contrib::Pg::Ext::TAG_DB_NAME)).to eq(dbname)
+          expect(span.get_tag(Datadog::Tracing::Metadata::Ext::TAG_COMPONENT))
+            .to eq(Datadog::Tracing::Contrib::Pg::Ext::TAG_COMPONENT)
+          expect(span.get_tag(Datadog::Tracing::Metadata::Ext::TAG_OPERATION))
+            .to eq(Datadog::Tracing::Contrib::Pg::Ext::TAG_OPERATION_QUERY)
+          expect(span.get_tag(Datadog::Tracing::Metadata::Ext::TAG_PEER_SERVICE))
+            .to eq(Datadog::Tracing::Contrib::Pg::Ext::DEFAULT_PEER_SERVICE_NAME)
+          expect(span.get_tag(Datadog::Tracing::Metadata::Ext::TAG_PEER_HOSTNAME)).to eq(host)
+          expect(span.get_tag(Datadog::Tracing::Metadata::Ext::DB::TAG_INSTANCE)).to eq(dbname)
+          expect(span.get_tag(Datadog::Tracing::Metadata::Ext::DB::TAG_USER)).to eq(user)
+          expect(span.get_tag(Datadog::Tracing::Metadata::Ext::DB::TAG_SYSTEM))
+            .to eq(Datadog::Tracing::Contrib::Pg::Ext::SPAN_SYSTEM)
+          expect(span.get_tag(Datadog::Tracing::Metadata::Ext::NET::TAG_TARGET_HOST)).to eq(host)
+          expect(span.get_tag(Datadog::Tracing::Metadata::Ext::NET::TAG_TARGET_PORT)).to eq(port.to_i)
+          expect(span.get_tag(Datadog::Tracing::Metadata::Ext::NET::TAG_DESTINATION_NAME)).to eq(host)
+          expect(span.get_tag(Datadog::Tracing::Metadata::Ext::NET::TAG_DESTINATION_PORT)).to eq(port.to_i)
+          expect(span.get_tag(Datadog::Tracing::Metadata::Ext::DB::TAG_ROW_COUNT)).to eq(1)
+        end
+
+        it_behaves_like 'analytics for integration' do
+          let(:analytics_enabled_var) { Datadog::Tracing::Contrib::Pg::Ext::ENV_ANALYTICS_ENABLED }
+          let(:analytics_sample_rate_var) { Datadog::Tracing::Contrib::Pg::Ext::ENV_ANALYTICS_SAMPLE_RATE }
+        end
+
+        it_behaves_like 'a peer service span' do
+          let(:peer_hostname) { host }
+        end
+
+        it_behaves_like 'measured span for integration', false
+      end
+
+      context 'when a failed query is made' do
+        before { expect { conn.sync_exec_prepared('invalid prepared select 1', ['INVALID']) }.to raise_error(PG::Error) }
+
+        it 'traces failed queries' do
+          expect(spans.count).to eq(1)
+          expect(span.status).to eq(1)
+          expect(span.get_tag('error.msg'))
+            .to include('ERROR') & include('prepared statement "invalid prepared select 1" does not exist')
```

Comment:
We noticed we were verifying these error statuses and tags very frequently, so we created a few helpers. Here's an example on how to use them:
https://github.com/DataDog/dd-trace-rb/blob/8fcf8bf5615fdd96d33d7ef8000fd2fac589c1cf/spec/datadog/tracing/contrib/grpc/datadog_interceptor/server_spec.rb#L75-L78

If you don't provide an exact object to match, it will simply check for the existence of that tag. For example:
```ruby
expect(span).to have_error_stack
```
Will fail, unless there is a non-blank string set as the `error.stack` tag.

You can use these compound matchers as well:
```ruby
expect(span).to have_error_message(include('ERROR') & include('prepared statement "invalid prepared select 1" does not exist'))
```


Here are the helper declarations, for reference: https://github.com/DataDog/dd-trace-rb/blob/master/spec/support/span_helpers.rb#L6-L57. Here's the RSpec docs about custom matchers in general: https://relishapp.com/rspec/rspec-expectations/docs/custom-matchers/define-a-custom-matcher

Reply:
@jennchenn, `expect(span.status).to eq(1)` can also be replaced with `expect(span).to have_error` (or `expect(span).to_not have_error` if you want to verify the opposite condition).